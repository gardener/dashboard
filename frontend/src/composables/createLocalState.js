//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import { effectScope } from 'vue'

export function createLocalState (stateFactory) {
  let state
  const scope = effectScope(true)

  return (...args) => {
    state = scope.run(() => stateFactory(...args))
    return state
  }
}
