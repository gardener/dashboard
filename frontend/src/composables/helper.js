//
// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  isEmpty,
  isNil,
  isObject,
} from '@/lodash'

export function cleanup (obj) {
  const cleanupObject = obj => {
    const cleanObj = {}
    for (const [key, value] of Object.entries(obj)) {
      const cleanValue = cleanupValue(value)
      if (cleanValue !== null) {
        cleanObj[key] = cleanValue
      }
    }
    return cleanObj
  }

  const cleanupArray = values => {
    const cleanValues = []
    for (const value of values) {
      const cleanValue = cleanupValue(value)
      if (cleanValue !== null) {
        cleanValues.push(cleanValue)
      }
    }
    return cleanValues
  }

  const cleanupValue = value => {
    if (isNil(value)) {
      return null
    }
    if (!isObject(value)) {
      return value
    }
    if (isEmpty(value)) {
      return null
    }
    const obj = Array.isArray(value)
      ? cleanupArray(value)
      : cleanupObject(value)
    if (isEmpty(obj)) {
      return null
    }
    return obj
  }

  return cleanupValue(obj)
}
