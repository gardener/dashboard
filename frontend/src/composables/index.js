//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

export { useApi } from './useApi'
export { useAuthn } from './useAuthn'
export { useLogger } from './useLogger'
export { useNav } from './useNav'
export { useTheme } from './useTheme'
export { useToken } from './useToken'
export { useUser } from './useUser'
