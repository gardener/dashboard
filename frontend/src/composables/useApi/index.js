//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import { markRaw } from 'vue'

import api from './api'
import { registry } from './fetch'

export const useApi = () => {
  return markRaw(api)
}

export const useInterceptors = () => {
  return markRaw(registry)
}
