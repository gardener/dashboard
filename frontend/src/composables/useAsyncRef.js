//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  ref,
  watch,
} from 'vue'

import { useLogger } from '@/composables/useLogger'

import pTimeout from '@/utils/p-timeout'

export const useAsyncRef = (key, options = {}) => {
  const {
    logger = useLogger(),
  } = options
  let keyRefPromise

  const keyRef = ref(null)

  const fulfiller = {
    resolve: () => {},
  }

  const init = () => {
    keyRefPromise = new Promise(resolve => {
      fulfiller.resolve = resolve
    })
  }

  init()

  watch(keyRef, value => {
    if (value) {
      fulfiller.resolve(value)
      fulfiller.resolve = () => {}
    } else {
      init()
    }
  })

  return {
    [key + 'Ref']: keyRef,
    [key]: {
      vm: ({ timeout = 3000 } = {}) => {
        return pTimeout(keyRefPromise, timeout, `async reference ${key} timed out`)
      },
      async dispatch (obj, ...args) {
        const { method, ...options } = typeof obj === 'string'
          ? { method: obj }
          : obj
        try {
          return (await this.vm(options))[method](...args)
        } catch (err) {
          logger.error(err.message)
        }
      },
      async get (name, options) {
        try {
          return (await this.vm(options))[name]
        } catch (err) {
          logger.error(err.message)
        }
      },
    },
  }
}
