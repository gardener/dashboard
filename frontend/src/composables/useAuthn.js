//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import { createSharedComposable, useBrowserLocation } from '@vueuse/core'
import { useToken, useLogger } from '@/composables'

// Local
import {
  createError,
  createNoUserError,
  createClockSkewError,
  isUnauthorizedError,
  isNoUserError,
  isClockSkewError,
} from '@/utils/errors'

function delay (duration = 0) {
  return new Promise(resolve => setTimeout(resolve, duration))
}

export const useAuthn = createSharedComposable(() => {
  const logger = useLogger()
  const location = useBrowserLocation()
  const token = useToken()

  let refreshTokenPromise

  async function createTokenRefreshRequest () {
    const timestamp = Math.floor(Date.now() / 1000)
    const response = await fetch('/auth/token', {
      method: 'POST',
      cache: 'no-store',
      headers: {
        accept: 'application/json',
        'x-requested-with': 'XMLHttpRequest',
        'content-type': 'application/json',
      },
      body: JSON.stringify({ timestamp }),
    })

    const statusCode = response.status
    if (statusCode >= 200 && statusCode < 300) {
      if (token.isEmpty()) {
        throw createNoUserError()
      }
      if (token.isRefreshRequired()) {
        throw createClockSkewError()
      }
    }

    if (statusCode >= 400) {
      let message = `Token refresh failed with status code ${statusCode}`
      try {
        const data = await response.json()
        Object.defineProperty(response, 'data', { value: data })
        if (data.message) {
          message = data.message
          if (statusCode === 401) {
            /*
             * The OPError message has the following format `${error} (${error_description})`
             * (see https://github.com/panva/node-openid-client/blob/1e3892e6222fdd1956735f97584ebc722fcebdd3/lib/errors.js#L5)
             * The error and the error_description are values returned in the error response
             * from the the OpenID Connect Provider (OP). We use the original OP error_description
             * as error message if possible.
             */
            const matches = /^(.+) \((.+)\)$/.exec(message)
            if (matches && matches.length > 2) {
              message = matches[2]
            }
          }
        }
      } catch (err) { /* ignore error */ }
      throw createError(statusCode, message, { response })
    }

    throw createError(statusCode, 'Token refresh failed', { response })
  }

  function getUser () {
    return token.payload.value
  }

  async function refreshToken () {
    try {
      if (token.isEmpty()) {
        throw createNoUserError()
      }
      if (!token.isRefreshRequired()) {
        return
      }
      const user = getUser()
      logger.debug('Acquiring token refresh lock (%s)', user.rti)
      await navigator.locks.request('token-refresh-request', async () => {
        if (token.isEmpty()) {
          throw createNoUserError()
        }
        if (!token.isRefreshRequired()) {
          return
        }
        const oldUser = getUser()
        logger.debug('Refreshing token (%s)', oldUser.rti)
        await createTokenRefreshRequest()
        const user = getUser()
        logger.debug('Token has been refreshed (%s <- %s)', user.rti, oldUser.rti)
      })
    } catch (err) {
      logger.error('Token refresh failed: %s - %s', err.name, err.message)
      let frameRequestCallback
      if (isNoUserError(err)) {
        frameRequestCallback = () => signin()
      } else if (isUnauthorizedError(err) || isClockSkewError(err)) {
        frameRequestCallback = () => signout(err)
      }
      if (typeof frameRequestCallback === 'function') {
        window.requestAnimationFrame(frameRequestCallback)
        // delay the error by 500 ms to avoid rendering or navigation before redirection has started
        await delay(500)
      }
      throw err
    }
  }

  function signout (err) {
    token.remove()
    const url = new URL('/auth/logout', location.value.origin)
    if (err) {
      url.searchParams.set('error[name]', err.name)
      url.searchParams.set('error[message]', err.message)
    }
    location.value.href = url
  }

  function signin () {
    const url = new URL('/login', location.value.origin)
    location.value.href = url
  }

  function signinWithOidc (redirectPath = '/') {
    const url = new URL('/auth', location.value.origin)
    const redirectUrl = new URL(redirectPath, location.value.origin)
    url.searchParams.set('redirectUrl', redirectUrl)
    location.value.href = url
  }

  function ensureValidToken () {
    if (!refreshTokenPromise) {
      refreshTokenPromise = refreshToken(token.payload).finally(() => {
        refreshTokenPromise = undefined
      })
    }
    return refreshTokenPromise
  }

  function isSessionExpired () {
    return token.isExpired()
  }

  return {
    signin,
    signout,
    signinWithOidc,
    ensureValidToken,
    isSessionExpired,
  }
})
