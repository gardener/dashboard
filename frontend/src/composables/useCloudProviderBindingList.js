//
// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import { computed } from 'vue'

import { useCredentialStore } from '@/store/credential'

import { isSharedCredential } from '@/utils'

import filter from 'lodash/filter'

export const useCloudProviderBindingList = (providerType, options = {}) => {
  const {
    credentialStore = useCredentialStore(),
  } = options

  return computed(() => {
    return filter(credentialStore.cloudProviderBindingList, binding => {
      if (binding.provider?.type !== providerType.value) {
        return false
      }
      if (binding._isDnsBinding) {
        return !isSharedCredential(binding) // shared dns secret not supported by dns extension
      }
      return true
    })
  })
}
