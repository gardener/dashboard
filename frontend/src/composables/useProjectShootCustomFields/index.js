//
// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import { computed } from 'vue'

import { useLogger } from '@/composables/useLogger'
import { useProjectMetadata } from '@/composables/useProjectMetadata'

import {
  cloneDeep,
  filter,
  isEmpty,
  isObject,
  map,
  some,
  camelCase,
  find,
} from '@/lodash'

export function useProjectShootCustomFields (projectItem, options = {}) {
  const {
    logger = useLogger(),
  } = options

  const {
    projectName,
    getProjectAnnotation,
    setProjectAnnotation,
    unsetProjectAnnotation,
  } = useProjectMetadata(projectItem)

  const rawShootCustomFields = computed({
    get () {
      const shootCustomFields = getProjectAnnotation('dashboard.gardener.cloud/shootCustomFields')
      if (!shootCustomFields) {
        return null
      }

      try {
        return JSON.parse(shootCustomFields)
      } catch (error) {
        logger.error('could not parse custom fields', error.message)
        return null
      }
    },
    set (value) {
      if (isEmpty(value)) {
        unsetProjectAnnotation('dashboard.gardener.cloud/shootCustomFields')
      } else {
        setProjectAnnotation('dashboard.gardener.cloud/shootCustomFields', JSON.stringify(value))
      }
    },
  })

  const shootCustomFields = computed(() => {
    if (!rawShootCustomFields.value) {
      return []
    }
    let shootCustomFields = rawShootCustomFields.value

    // extract values from legacy object format
    if (typeof rawShootCustomFields.value === 'object') {
      shootCustomFields = Object.values(rawShootCustomFields.value)
    }

    // filter out invalid custom fields
    shootCustomFields = filter(shootCustomFields, customField => {
      if (isEmpty(customField)) {
        return false // omit null values
      }
      if (some(customField, isObject)) {
        return false // omit custom fields with object values
      }
      return customField.name && customField.path
    })

    const defaultProperties = {
      showColumn: true,
      columnSelectedByDefault: true,
      showDetails: true,
      sortable: true,
      searchable: true,
    }
    shootCustomFields = map(shootCustomFields, customField => {
      const key = generateKeyFromName(customField.name)
      const customFieldWithKey = {
        ...defaultProperties,
        ...customField,
        weight: Number(customField.weight) || 0,
      }
      Object.defineProperty(customFieldWithKey, 'key', {
        value: key,
        enumerable: false, // Ensure the key is not included in JSON.stringify()
      })
      return customFieldWithKey
    })
    return shootCustomFields
  })

  function addShootCustomField (customField) {
    shootCustomFields.value.push(cloneDeep(customField))
    rawShootCustomFields.value = shootCustomFields.value
  }

  function deleteShootCustomField (customField) {
    const key = generateKeyFromName(customField.name)
    const index = shootCustomFields.value.findIndex(field => field.key === key)
    if (index !== -1) {
      shootCustomFields.value.splice(index, 1)
    }
    rawShootCustomFields.value = shootCustomFields.value
  }

  function replaceShootCustomField (oldCustomField, newCustomField) {
    const key = generateKeyFromName(oldCustomField.name)
    const index = shootCustomFields.value.findIndex(field => field.key === key)
    if (index !== -1) {
      shootCustomFields.value.splice(index, 1, cloneDeep(newCustomField))
    }
    rawShootCustomFields.value = shootCustomFields.value
  }

  function isShootCustomFieldNameUnique (name) {
    for (const customField of shootCustomFields.value) {
      if (customField.name === name) {
        return false
      }

      // Ensure uniqueness when used as custom header
      if (generateKeyFromName(customField.name) === generateKeyFromName(name)) {
        return false
      }
    }

    return true
  }

  function getShootCustomFieldsPatchDocument () {
    return {
      metadata: {
        name: projectName.value,
        annotations: {
          'dashboard.gardener.cloud/shootCustomFields': getProjectAnnotation('dashboard.gardener.cloud/shootCustomFields', null),
        },
      },
    }
  }

  function getCustomFieldByKey (key) {
    return find(shootCustomFields.value, ['key', key])
  }

  function generateKeyFromName (name) {
    return `Z_${camelCase(name)}`
  }

  return {
    shootCustomFields,
    rawShootCustomFields,
    addShootCustomField,
    deleteShootCustomField,
    replaceShootCustomField,
    isShootCustomFieldNameUnique,
    getShootCustomFieldsPatchDocument,
    getCustomFieldByKey,
    generateKeyFromName,
  }
}
