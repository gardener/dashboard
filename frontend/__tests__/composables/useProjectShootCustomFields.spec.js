//
// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  ref,
  reactive,
} from 'vue'
import {
  describe,
  it,
  expect,
  vi,
  beforeEach,
} from 'vitest'

import { useProjectShootCustomFields } from '@/composables/useProjectShootCustomFields'
import {
  formatValue,
  isCustomField,
} from '@/composables/useProjectShootCustomFields/helper'

// Mock dependencies
vi.mock('@/composables/useLogger', () => ({
  useLogger: () => ({
    error: vi.fn(),
  }),
}))

let annotations

beforeEach(() => {
  annotations = reactive({})
})

vi.mock('@/composables/useProjectMetadata', () => ({
  useProjectMetadata: projectItem => ({
    getProjectAnnotation: vi.fn(key => annotations[key] || null),
    setProjectAnnotation: vi.fn((key, value) => {
      annotations[key] = value
    }),
    unsetProjectAnnotation: vi.fn(key => {
      delete annotations[key]
    }),
  }),
}))

describe('useProjectShootCustomFields', () => {
  let projectItem
  let options

  beforeEach(() => {
    projectItem = ref({})
    options = {}
  })

  it('should initialize shootCustomFields correctly with legacy data', () => {
    annotations['dashboard.gardener.cloud/shootCustomFields'] = JSON.stringify({
      shootStatus: {
        name: 'Shoot Status',
        path: 'metadata.labels["shoot.gardener.cloud/status"]',
        icon: 'mdi-heart-pulse',
        tooltip: 'Indicates the health status of the cluster',
        defaultValue: 'unknown',
        showColumn: true,
        columnSelectedByDefault: true,
        weight: 950,
        searchable: true,
        sortable: true,
        showDetails: true,
      },
      networking: {
        name: 'Networking Type',
        path: 'spec.networking.type',
        icon: 'mdi-table-network',
        showColumn: false,
      },
    })
    const { shootCustomFields } = useProjectShootCustomFields(projectItem, options)
    expect(shootCustomFields.value.length).toBe(2)
    expect(shootCustomFields.value).toMatchSnapshot()
  })

  it('should add a custom field', async () => {
    const { shootCustomFields, addShootCustomField } = useProjectShootCustomFields(projectItem, options)
    const customField = { name: 'Field1', path: 'path1' }
    expect(shootCustomFields.value.length).toBe(0)
    addShootCustomField(customField)
    expect(shootCustomFields.value.length).toBe(1)
    expect(shootCustomFields.value).toMatchSnapshot()
  })

  it('should delete a custom field', async () => {
    const { shootCustomFields, addShootCustomField, deleteShootCustomField } = useProjectShootCustomFields(projectItem, options)
    const customField = { name: 'Field1', path: 'path1' }
    addShootCustomField(customField)
    expect(shootCustomFields.value.length).toBe(1)
    deleteShootCustomField(customField)
    expect(shootCustomFields.value.length).toBe(0)
  })

  it('should replace a custom field', async () => {
    const { shootCustomFields, addShootCustomField, replaceShootCustomField } = useProjectShootCustomFields(projectItem, options)
    const oldCustomField = { name: 'Field1', path: 'path1' }
    const newCustomField = { name: 'Field2', path: 'path2' }
    addShootCustomField(oldCustomField)
    expect(shootCustomFields.value.length).toBe(1)
    replaceShootCustomField(oldCustomField, newCustomField)
    expect(shootCustomFields.value.length).toBe(1)
    expect(shootCustomFields.value).toContainEqual(expect.objectContaining(newCustomField))
    expect(shootCustomFields.value).not.toContainEqual(expect.objectContaining(oldCustomField))
  })

  it('should check if custom field name is unique', async () => {
    const { addShootCustomField, isShootCustomFieldNameUnique } = useProjectShootCustomFields(projectItem, options)
    const customField = { name: 'Field1', path: 'path1' }
    addShootCustomField(customField)
    expect(isShootCustomFieldNameUnique('Field1')).toBe(false)
    expect(isShootCustomFieldNameUnique('Field2')).toBe(true)
  })

  it('should get custom field by key', async () => {
    const { addShootCustomField, getCustomFieldByKey } = useProjectShootCustomFields(projectItem, options)
    const customField = { name: 'Field1', path: 'path1' }
    addShootCustomField(customField)
    addShootCustomField({ name: 'Field2', path: 'path2' })
    const key = 'Z_field1'
    const result = getCustomFieldByKey(key)
    expect(result).toEqual(expect.objectContaining(customField))
  })

  it('should handle empty custom fields', async () => {
    const { shootCustomFields, rawShootCustomFields } = useProjectShootCustomFields(projectItem, options)
    rawShootCustomFields.value = null
    expect(shootCustomFields.value).toEqual([])
  })

  it('should handle invalid JSON in custom fields', async () => {
    const { shootCustomFields, rawShootCustomFields } = useProjectShootCustomFields(projectItem, options)
    rawShootCustomFields.value = 'invalid JSON'
    expect(shootCustomFields.value).toEqual([])
  })

  it('should generate key from name correctly', () => {
    const { generateKeyFromName } = useProjectShootCustomFields(projectItem, options)
    const name = 'Custom Field Name'
    const key = generateKeyFromName(name)
    expect(key).toBe('Z_customFieldName')
  })

  it('should skip custom fields with invalid values', () => {
    annotations = reactive({
      'dashboard.gardener.cloud/shootCustomFields': JSON.stringify([
        { name: 'Valid Field', path: 'path1' },
        { name: 'Invalid Field1', path: 'path2', value: { key: 'value' } }, // ignored - no objects allowed as values of custom field properties
        { name: 'Invalid Field2', path: { foo: 'bar' } }, // no objects allowed as values of custom field properties
        { name: 'Invalid Field3' }, // ignored, missing required property path
        { path: 'path3' }, // ignored, missing required property name
        {}, // ignored
        null, // ignored
      ]),
    })

    const { shootCustomFields } = useProjectShootCustomFields(projectItem, options)
    expect(shootCustomFields.value.length).toBe(1)
    expect(shootCustomFields.value).toMatchSnapshot()
  })

  it('should ensure the key is not part of the rendered JSON.stringify annotation when adding a new entry', () => {
    const { shootCustomFields, addShootCustomField } = useProjectShootCustomFields(projectItem, options)
    const customField = { name: 'Field1', path: 'path1' }

    // Add the custom field
    addShootCustomField(customField)

    // Verify the key is set correctly
    const addedField = shootCustomFields.value.find(field => field.name === 'Field1')
    expect(addedField.key).toBe('Z_field1')

    // Verify the key is not part of the JSON stringified output in annotations
    const annotationString = annotations['dashboard.gardener.cloud/shootCustomFields']
    expect(annotationString).not.toContain('key')
    expect(annotationString).toMatchSnapshot()
  })

  describe('helper', () => {
    describe('formatValue', () => {
      it('should join array elements with a separator', () => {
        const result = formatValue(['a', 'b', 'c'], ',')
        expect(result).toBe('a,b,c')
      })

      it('should return the value as is for non-array values', () => {
        const result = formatValue('string', ',')
        expect(result).toBe('string')
      })

      it('should return undefined for object values', () => {
        const result = formatValue({ key: 'value' }, ',')
        expect(result).toBeUndefined()
      })
    })

    describe('isCustomField', () => {
      it('should return true for keys starting with "Z_"', () => {
        const result = isCustomField('Z_customField')
        expect(result).toBe(true)
      })

      it('should return false for keys not starting with "Z_"', () => {
        const result = isCustomField('customField')
        expect(result).toBe(false)
      })
    })
  })
})
