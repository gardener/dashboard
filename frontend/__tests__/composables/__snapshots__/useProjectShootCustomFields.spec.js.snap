// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useProjectShootCustomFields > should add a custom field 1`] = `
[
  {
    "columnSelectedByDefault": true,
    "name": "Field1",
    "path": "path1",
    "searchable": true,
    "showColumn": true,
    "showDetails": true,
    "sortable": true,
    "weight": 0,
  },
]
`;

exports[`useProjectShootCustomFields > should ensure the key is not part of the rendered JSON.stringify annotation when adding a new entry 1`] = `"[{"name":"Field1","path":"path1"}]"`;

exports[`useProjectShootCustomFields > should initialize shootCustomFields correctly with legacy data 1`] = `
[
  {
    "columnSelectedByDefault": true,
    "defaultValue": "unknown",
    "icon": "mdi-heart-pulse",
    "name": "Shoot Status",
    "path": "metadata.labels["shoot.gardener.cloud/status"]",
    "searchable": true,
    "showColumn": true,
    "showDetails": true,
    "sortable": true,
    "tooltip": "Indicates the health status of the cluster",
    "weight": 950,
  },
  {
    "columnSelectedByDefault": true,
    "icon": "mdi-table-network",
    "name": "Networking Type",
    "path": "spec.networking.type",
    "searchable": true,
    "showColumn": false,
    "showDetails": true,
    "sortable": true,
    "weight": 0,
  },
]
`;

exports[`useProjectShootCustomFields > should skip custom fields with invalid values 1`] = `
[
  {
    "columnSelectedByDefault": true,
    "name": "Valid Field",
    "path": "path1",
    "searchable": true,
    "showColumn": true,
    "showDetails": true,
    "sortable": true,
    "weight": 0,
  },
]
`;
